src/porepy/compositional/states.py

```diff
@@ -0,0 +1,364 @@
+
+    rho: np.ndarray = field(default_factory=lambda: np.zeros(0))
+    """Density."""
+
+    @property
+    def v(self) -> np.ndarray:
+        """Specific volume as the reciprocal of :attr:`rho`.
+
+        Returns zeros, where :attr:`rho` is zero.
+
+        """
+        v = np.zeros_like(self.rho)
+        # special treatment for zero values to avoid division-by zero errors
+        idx = self.rho > 0.0
+        v[idx] = 1.0 / self.rho[idx]
+        return v
+
+
+@dataclass
+class PhaseState(ExtensiveState):
+    """An extended state description for physical phases, including derivatives
+    of extensive properties and properties which are not state functions such
+    as viscosity and thermal conductivity.
+
+    Derivatives are denoted with a ``d*`` and the derivative values are stored
+    along the first axis of a 2D array (i.e. each rot represents the derivative with
+    respect to 1 dependency)
+
+    The state of a phase is additionally characterized by an integer representing the
+    phase type, and the values of fractions per component.
+
+    """
+
+    phasetype: int = 0
+    """Type of the phase. Defaults to 0 (liquid)."""
+
+    x: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Fractions for each component in a phase, stored row-wise per component
+    in a 2D array.
+
+    Fractions of components in a phase are relative to the fraction of a phase.
+
+    """
+
+    phis: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Fugacity coefficients per component in this phase, stored row-wise per component
+    in a 2D array."""
+
+    dh: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Derivatives of the specific enthalpy."""
+
+    drho: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Derivatives of the density."""
+
+    dphis: np.ndarray = field(default_factory=lambda: np.zeros((0, 0, 0)))
+    """Derivatives of fugacity coefficients.
+
+    This is a 3D array!
+
+    The first axis is associated with components in this phase and the second axis with
+    the derivatives. The third axis is for the values.
+
+    """
+
+    mu: np.ndarray = field(default_factory=lambda: np.zeros(0))
+    """Dynamic viscosity."""
+
+    dmu: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Derivatives of the dynamic viscosity."""
+
+    kappa: np.ndarray = field(default_factory=lambda: np.zeros(0))
+    """Thermal conductivity."""
+
+    dkappa: np.ndarray = field(default_factory=lambda: np.zeros((0, 0)))
+    """Derivatives of the thermal conductivity."""
+
+    @property
+    def dv(self) -> np.ndarray:
+        """Derivatives of the specific volume, expressed as the reciprocal of
+        density.
+
+        The chainrule is applied to compute ``dv`` from ``d(1/rho)``.
+
+        The derivatives are set to zero, where :attr:`rho` is zero.
+
+        """
+        # Treatment to avoid division by zero errors
+        idx = self.rho > 0.0
```

Comment:
It is not critical here, since in the subsurface we are not dealing with such tiny densities.

The ``eps`` (in ``utils.compute_saturations`` for example) on the other hand is required, because in the iterative sense it happens sometimes that fractions are ``1+- 1e-10`` or something like that. I don't think the program can safely decide what is *saturated* in this sense, so the user has to decide (though I provide a default value).